e2e/nomostest/portforwarder/port_forwarder.go

```diff
@@ -233,24 +214,95 @@ func (pf *PortForwarder) start() error {
 	// Restart the port forward if the process exits prematurely (e.g. due to pod eviction)
 	go func() {
 		for {
-			if err := pf.watcher.WatchForNotFound(kinds.Pod(), pf.pod, pf.ns, testwatcher.WatchTimeout(0), testwatcher.WatchContext(pf.ctx)); err != nil {
-				if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
-					// stop retrying, the subprocess will be killed with the context.
-					return
-				}
-				// Watch automatically retries, so this shouldn't happen unless
-				// there's a terminal error.
-				pf.logger.Infof("Watch exited, restarting port-forward for %s/%s: %v", pf.ns, pf.deployment, err)
-			}
-			if err := pf.portForwardToDeployment(true); err != nil {
-				pf.logger.Info(err)
-				// this is a potentially fatal error since we failed to restart the port
-				// forward. we log the error and return in case the test is already done
-				// with the port forward. the test will otherwise error the next time it
-				// tries to use the local port.
+			if err := pf.waitAndRestart(); err != nil {
+				pf.logger.Infof("PortForwarder finished: %v", err)
 				return
 			}
 		}
 	}()
 	return nil
 }
+
+// waitAndRestart will block until seeing an indicator that the port-forward is
+// no longer healthy. If that occurs, it will restart the port-forward
+func (pf *PortForwarder) waitAndRestart() error {
+	err := pf.waitForFailure()
+	if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
+		return fmt.Errorf("exiting - PortForwarder for pod %s/%s received done signal", pf.ns, pf.pod)
+	}
+	pf.logger.Infof("received signal for port-forward for %s/%s. err: %v", pf.ns, pf.pod, err)
+	if err := pf.portForwardToDeployment(true); err != nil {
+		// this is a potentially fatal error since we failed to restart the port
+		// forward. we log the error and return in case the test is already done
+		// with the port forward. the test will otherwise error the next time it
+		// tries to use the local port.
+		return err
+	}
+	return nil
+}
+
+// waitForFailure waits for any failure condition which indicates the port-forward
+// must be restarted. The port-forward must be restarted if:
+// - the subprocess is killed
+// - the pod is restarted
+func (pf *PortForwarder) waitForFailure() error {
+	ctx, cancel := context.WithCancel(pf.ctx)
+	cmdChan := functionWithChannel(func() error {
+		// cmd.Wait does not accept a context, so create a wrapper that cancels with the ctx
+		cmdWaitChan := functionWithChannel(pf.cmd.Wait)
+		var chanErr error
+		select {
+		case <-ctx.Done():
+			chanErr = ctx.Err()
+		case err := <-cmdWaitChan:
+			chanErr = err
+		}
+		// kill the subprocess. This should force the cmd.Wait goroutine to return, if it hasn't already.
+		if err := pf.cmd.Process.Kill(); err != nil && !errors.Is(err, os.ErrProcessDone) {
+			// potentially bad if this fails, cmd.Wait may never return?
```

Comment:
This should probably return the kill error. If we fail to kill a process we need to stop the tests. If we exit the test process, it won't matter that the wait was still running in a goroutine.

Reply:
The question is - if this fails, do we still wait for the goroutine to exit? AFAICT this would be an unexpected error which means we may not be able to kill the subprocess, which means we may be waiting forever.

If we do wait for `cmd.Wait` to exit, this means the subprocess did exit and we probably want to return the original `chanErr` so that we can proceed as normal.

If we don't wait for `cmd.Wait` to exit, then we could return the `Kill` error but we may be leaking the goroutine

We could potentially do something like `panic` or `Fatal`. WDYT?